skbio/metadata/_interval.py

```diff
@@ -0,0 +1,312 @@
+
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in inv_md.intervals:
+            if loc is not None:
+                start, end = loc
+                self._intervals.add(start, end, inv_md)
+
+        self._metadata.append(inv_md)
+
+    def _rebuild_tree(self, intervals):
+        self._intervals = IntervalTree()
+        for f in intervals:
+            for inv in f.intervals:
+                start, end = inv
+                self._intervals.add(start, end, f)
+
+    def _query_interval(self, interval):
+        start, end = _polish_interval(interval)
+        invs = self._intervals.find(start, end)
+        return invs
+
+    def _query_attribute(self, intervals, metadata):
+        if metadata is None:
+            return []
+
+        queries = []
+        for inv in intervals:
+            for (key, value) in metadata.items():
+                if inv[key] != value:
+                    continue
+                queries.append(inv)
+        return queries
+
+    def query(self, intervals=None, boundaries=None, metadata=None):
+        """ Looks up Interval objects with the intervals, boundaries and keywords.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object.
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object.
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object.
+
+        Returns
+        -------
+        list, Interval
+            A list of Intervals satisfying the search criteria.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> interval_metadata = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+        ...                       boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+        ...                       boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        [Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})]
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval.
+        2. Query by key/val pair (i.e. gene=sagA).
+
+        """
+        if self._is_stale_tree:
+            self._rebuild_tree(self._metadata)
+            self._is_stale_tree = False
+
+        invs = set()
+
+        # Find queries by interval
+        if intervals is not None:
+            for value in intervals:
+                invs.update(self._query_interval(value))
+
+        # Find queries by feature attribute
+        if len(invs) == 0 and metadata is not None:
+            invs = set(self._metadata)
+
+        if metadata is not None:
+            invs = self._query_attribute(list(invs), metadata)
```

Comment:
Done

On Thu, May 19, 2016 at 11:11 PM, Zech Xu <notifications@github.com> wrote:

> In skbio/metadata/_interval.py
> <https://github.com/biocore/scikit-bio/pull/1356#discussion_r63995174>:
>
> > +            self._rebuild_tree(self._metadata)
> > +            self._is_stale_tree = False
> > +
> > +        invs = set()
> > +
> > +        # Find queries by interval
> > +        if intervals is not None:
> > +            for value in intervals:
> > +                invs.update(self._query_interval(value))
> > +
> > +        # Find queries by feature attribute
> > +        if len(invs) == 0 and metadata is not None:
> > +            invs = set(self._metadata)
> > +
> > +        if metadata is not None:
> > +            invs = self._query_attribute(list(invs), metadata)
>
> no need to convert to list
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1356/files/4767f8e7b6de74b0308879ee2fc2b7da9dceb048#r63995174>
>